cmd/agola/cmd/logget.go

```diff
@@ -0,0 +1,137 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+
+	gwapitypes "agola.io/agola/services/gateway/api/types"
+	gwclient "agola.io/agola/services/gateway/client"
+
+	"github.com/spf13/cobra"
+	errors "golang.org/x/xerrors"
+)
+
+var cmdLogGet = &cobra.Command{
+	Use:   "get",
+	Short: "get a setup/step log",
+	Run: func(cmd *cobra.Command, args []string) {
+		if err := logGet(cmd, args); err != nil {
+			log.Fatalf("err: %v", err)
+		}
+	},
+}
+
+type logGetOptions struct {
+	runid    string
+	taskname string
+	step     int
+	setup    bool
+	follow   bool
+	output   string
+}
+
+var logGetOpts logGetOptions
+
+func init() {
+	flags := cmdLogGet.Flags()
+
+	flags.StringVar(&logGetOpts.runid, "runid", "", "RunId")
+	flags.StringVar(&logGetOpts.taskname, "taskname", "", "Task name")
+	flags.IntVar(&logGetOpts.step, "step", 0, "Step number")
+	flags.BoolVar(&logGetOpts.setup, "setup", false, "Setup step")
+	flags.BoolVar(&logGetOpts.follow, "follow", false, "Follow log stream")
+	flags.StringVar(&logGetOpts.output, "output", "", "Write output to file")
+
+	if err := cmdLogGet.MarkFlagRequired("runid"); err != nil {
+		log.Fatal(err)
+	}
+	if err := cmdLogGet.MarkFlagRequired("taskname"); err != nil {
+		log.Fatal(err)
+	}
+
+	cmdLog.AddCommand(cmdLogGet)
+}
+
+func logGet(cmd *cobra.Command, args []string) error {
+
+	flags := cmd.Flags()
+
+	if flags.Changed("step") && flags.Changed("setup") {
+		return errors.Errorf(`only one of "--step" or "--setup" can be provided`)
+	} else {
+		if !flags.Changed("step") && !flags.Changed("setup") {
+			return errors.Errorf(`one of "--step" or "--setup" must be provided`)
+		} else {
+			if logGetOpts.step < 0 {
+				return errors.Errorf("step number %d is invalid, it must be equal or greater than zero", logGetOpts.step)
+			}
+		}
+	}
+	if flags.Changed("follow") && flags.Changed("output") {
+		return errors.Errorf(`only one of "--follow" or "--output" can be provided`)
+	}
+
+	gwclient := gwclient.NewClient(gatewayURL, token)
+
+	run, _, err := gwclient.GetRun(context.TODO(), logGetOpts.runid)
+	if err != nil {
+		return errors.Errorf("runid %q not found", logGetOpts.runid)
+	}
+
+	var task *gwapitypes.RunResponseTask
+	var taskfound bool
+	for _, t := range run.Tasks {
+		if t.Name == logGetOpts.taskname {
+			task = t
+			taskfound = true
+			break
+		}
+	}
+	if !taskfound {
+		return errors.Errorf("task %q not found in run %q", logGetOpts.taskname, logGetOpts.runid)
+	}
+	log.Infof("getting log")
+	resp, err := gwclient.GetLogs(context.TODO(), logGetOpts.runid, task.ID, logGetOpts.setup, logGetOpts.step, logGetOpts.follow)
+	if err != nil {
+		return errors.Errorf("failed to get log: %v", err)
+	}
+	defer resp.Body.Close()
+
+	if flags.Changed("follow") {
+		if _, err := io.Copy(os.Stdout, resp.Body); err != nil {
```

Comment:
You can do the same `io.Copy` with or without `follow` when the output is stdout.